### PR TITLE
gh-123 Add task state diagram

### PR DESCRIPTION
**Pull request by veod32**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -323,7 +328,7 @@ queue = require 'queue'
 The request "require 'queue'" causes automatic creation of
 the `_queue` space, unless it already exists. The same request
 also sets the necessary space triggers and other objects
-associated with queues.
```
Comment (LeonidVas) [inline README.md:326]: Why did you decide to remove the double space ("\n" in md) here and below?
These are unnecessary changes. If you want to update something that not related with patchset, create a separate PR.
I suggest you read https://github.com/tarantool/tarantool/wiki/Code-review-procedure#general-coding-points-to-check and https://github.com/tarantool/tarantool/wiki/Code-review-procedure#commit-message .

File: README.md [hunk 2]
```diff
@@ -295,14 +295,19 @@ The `task_state` field takes one of the following values
...
+* '-' - the task has been **executed (done)** (a task is removed from the queue
+after it has been executed, so this value will rarely be seen)
+* '!' - the task is **buried** (disabled temporarily until further changes)
+* '~' - the task is **delayed** for some time.
+
+For better understanding of the task life cycle, below is the task state
+diagram with indication of all transitions between the tasks states mentioned
+above and triggers of these transitions.
+
+![Task state diagram](images/task_state_diag.png)
```
Comment (LeonidVas) [inline README.md:310]: It might be better to do it in a separate section (#Statechart). I don't know how it would look better. What do you think?

**Review by LeonidVas — CHANGES_REQUESTED** [review]
Hi! Thank you for the patchset.
I experimented with the diagram and it looks like I was able to achieve a nicer picture.
```
@startuml
[*]     -down->  READY   : put()
[*]     -down->  DELAYED : put('my_task_data',\n{ delay = delay })
READY   -down->  TAKEN   : take()
READY   -down->  DONE    : delete() /\nttl timeout
READY   -down->  BURIED  : bury()
TAKEN   -up->    READY   : release() /\nttr timeout
TAKEN   -left->  DELAYED : release\n(id, { delay = delay })
TAKEN   -down->  DONE    : ack() / delete()
TAKEN   -right-> BURIED  : bury()
BURIED  -down->  DONE    : delete() /\nttl timeout
BURIED  -up->    READY   : kick()
DELAYED -up->    READY   : timeout
DELAYED -down->  DONE    : delete()
DONE    -down->  [*]     : remove executed task
@enduml
```
state_diagram
I suggest adding images to the `doc/images` with the plantUML source for future editing.

File: README.md (hunk 2 above)
Comment (Totktonada) [inline README.md:310]: Please, save source text of the diagram within the repository and give pointers how to render the image.

Possible directory structure:

```
+ doc/
  + images/
    - statechart.png
    + src/
      - statechart.puml
      - README.md
```

I propose to describe steps to generate `doc/images/statechart.png` in `doc/images/src/README.md` or right within `doc/images/src/statechart.puml` in a comment if possible. Don't forget to include URL to the webpage if the process involves usage of an online tool.

Comment (Totktonada) [reply README.md:310]: Alternative to saving .png to the repository is using of an external rendering service: https://blog.anoff.io/2018-07-31-diagrams-with-plantuml/#github-integration

Comment (veod32) [reply README.md]: The alternative way you mentioned works. I think we should use it because it's more convenient: you just need to keep the source plantuml code in `/doc/scr/` and include a link to the PlantUML rendering service in readme.md via the standard markdown syntax ``.
To update the diagram, you need to change just the source plantuml code.
If you think we still need a kind of readme for this case, I will add it.

Comment (Totktonada) [reply README.md]: Nope, if we'll use this way, everything should just work and doc/src/README.md is not necessary.

Comment (veod32) [reply README.md]: Put the diagram in a separate section `Task state diagram`. Also, added reference links to the sections of the methods that trigger state transitions.

File: README.md (hunk 1 above)
Comment (veod32) [reply README.md:326]: I got your point on including only relevant updates into PR.

If these double spaces serve some purpose and are not just typos, I will correct this.

Comment (Totktonada) [reply README.md:326]: Two whitespaces at the end of line [issues](https://daringfireball.net/projects/markdown/syntax#p) a `<br />` tag into HTML. Personally I consider this feature as broken in two ways: the syntax is weird, the feature is not needed (it always better to just split text with paragraphs or use three-backticks code blocks). So I prefer to don't use it. And GFM (GitHub's Markdown dialect) anyway preserves line breaks, so it is useless on GitHub.

Anyway, Leonid is right, it is not relevant to the problem that you are solving.

File: README.md (hunk 2 above)
Comment (NickVolynkin) [reply README.md]: @Totktonada I've added a GH Action to (re)generate and save diagrams. Please have a look.

**Comment by NickVolynkin** [comment]
Oops, I've accidentally committed contents of `.idea`.

File: README.md (hunk 2 above)
Comment (Totktonada) [reply README.md]: This way we store artifacts on our side and don't depend on reliability of an external service. That's good.

There are several downsides:

* Unclear workflow: an extra commit will be added into the PR, but then one more extra commit will be added into `master`?
* A generated file in the repository.

The first one disquiets me, to be honest. I don't want touching of my branch (and, moreover, `master`) by a robot. Version control systems are for humans.

Since @LeonidVas maintaining the repository, this will be his decision.

**Review by LeonidVas — APPROVED** [review]
Hi!
I'll comfortable with one of https://github.com/tarantool/queue/pull/145#discussion_r553207873 , https://github.com/tarantool/queue/pull/145#discussion_r553206221 .
Please correct the code according to the comments, squash the commits, self-review the code and update the PR.

**Review by LeonidVas — CHANGES_REQUESTED** [review]
Mistakenly chose "Approve" instead of "Request changes" in https://github.com/tarantool/queue/pull/145#pullrequestreview-608510924

File: README.md (hunk 2 above)
Comment (veod32) [reply README.md]: OK. Decided to use the option of saving the diagram file (png or svg) in the repository [#145 (comment)](https://github.com/tarantool/queue/pull/145#discussion_r553206221). The option of using some external rendering service seems not very reliable.

Comment (veod32) [reply README.md]: Put the diagram here: ./doc/image/statediagram.svg
Wrote a short instruction on rendering diagram in SVG using https://www.planttext.com/. Put the instruction in statediagram.pulm as comments as Alexander asked.

**Review by LeonidVas — APPROVED** [review]

**Comment by Totktonada** [comment]
Does not look well in the dark mode.

image

**Comment by LeonidVas** [comment]
> Does not look well in the dark mode.
> 
> image

Let's do it beautifully in the future
